File: JavascriptAlgorithm/Recursive/subsets-078.js
var subsets = function (nums) {
  let res = [];
  let len = nums.length;
  const recursive = (curr, temp) => {
    if (curr === len) {
      res.push([...temp]);
      return;
    }
    temp.push(nums[curr]);
    recursive(curr + 1, temp);
    temp.pop();
    recursive(curr + 1, temp);
  };
  recursive(0, []);
  return res;
};
export { subsets };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter08/subsets_078.java
package algo.tzashinorpu.SecondRound.Chapter08;

import java.util.LinkedList;
import java.util.List;

public class subsets_078 {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        LinkedList<Integer> temp = new LinkedList<>();
        recursive(0, nums, temp, res);
        return res;
    }

    private void recursive(int level, int[] nums, LinkedList<Integer> temp, List<List<Integer>> res) {
        if (level == nums.length) {
            res.add(new LinkedList<>(temp));
            return;
        }
        temp.push(nums[level]);
        recursive(level + 1, nums, temp, res);
        temp.pop();
        recursive(level + 1, nums, temp, res);
    }
}
